File: python/benchmark_simplex.py
import random
import time
import psutil
import os
import json
import statistics
import gc

def mostrar_tabela(tabela):
    print("\nTabela Simplex:")
    for i, linha in enumerate(tabela):
        if i == len(tabela) - 1:
            print("Z:", end=" ")
        else:
            print(f"R{i+1}:", end=" ")
        print(["{:+8.2f}".format(x) for x in linha])
    print()

def encontrar_coluna_pivo(tabela):
    """Encontra a coluna com o menor valor negativo na linha objetivo"""
    ultima_linha = tabela[-1][:-1]
    valor_min = min(ultima_linha)
    if valor_min >= 0:
        return -1
    return ultima_linha.index(valor_min)

def encontrar_linha_pivo(tabela, coluna_pivo):
    """Teste da razão mínima para encontrar a linha pivô"""
    menor_razao = float('inf')
    linha_pivo = -1
    
    for i in range(len(tabela) - 1):
        elemento_pivo = tabela[i][coluna_pivo]
        rhs = tabela[i][-1]
        
        if elemento_pivo > 0:
            razao = rhs / elemento_pivo
            if razao >= 0 and razao < menor_razao:
                menor_razao = razao
                linha_pivo = i
    
    return linha_pivo

def pivotear(tabela, linha_pivo, coluna_pivo):
    """Executa as operações de pivoteamento"""
    num_linhas = len(tabela)
    num_colunas = len(tabela[0])
    pivo = tabela[linha_pivo][coluna_pivo]
    
    for j in range(num_colunas):
        tabela[linha_pivo][j] /= pivo
    
    for i in range(num_linhas):
        if i != linha_pivo:
            multiplicador = tabela[i][coluna_pivo]
            for j in range(num_colunas):
                tabela[i][j] -= multiplicador * tabela[linha_pivo][j]

def simplex(tabela, verbose=False, max_iteracoes=1000000):
    """Algoritmo Simplex padrão - retorna número de iterações"""
    iteracao = 0
    
    while iteracao < max_iteracoes:
        iteracao += 1
        
        coluna_pivo = encontrar_coluna_pivo(tabela)
        if coluna_pivo == -1:
            return iteracao  # Retorna número de iterações
        
        linha_pivo = encontrar_linha_pivo(tabela, coluna_pivo)
        if linha_pivo == -1:
            print("Problema ilimitado")
            return -1
        
        pivotear(tabela, linha_pivo, coluna_pivo)
    
    print(f"ATENÇÃO: Limite de {max_iteracoes} iterações atingido!")
    return iteracao

def construir_tabela_transporte(oferta, demanda, custos):
    """Constrói a tabela simplex para o problema de transporte"""
    m = len(oferta)
    n = len(demanda)
    num_vars = m * n
    
    total_oferta = sum(oferta)
    total_demanda = sum(demanda)
    if abs(total_oferta - total_demanda) > 1e-6:
        print(f"ERRO: Problema desbalanceado!")
        return None
    
    tabela = []
    
    # Restrições de oferta
    for i in range(m):
        linha = [0.0] * num_vars
        for j in range(n):
            linha[i * n + j] = 1.0
        
        folgas_oferta = [0.0] * m
        folgas_oferta[i] = 1.0
        folgas_demanda = [0.0] * n
        
        linha += folgas_oferta + folgas_demanda
        linha.append(float(oferta[i]))
        tabela.append(linha)
    
    # Restrições de demanda
    for j in range(n):
        linha = [0.0] * num_vars
        for i in range(m):
            linha[i * n + j] = 1.0
        
        folgas_oferta = [0.0] * m
        folgas_demanda = [0.0] * n
        folgas_demanda[j] = 1.0
        
        linha += folgas_oferta + folgas_demanda
        linha.append(float(demanda[j]))
        tabela.append(linha)
    
    # Função objetivo
    linha_obj = []
    for i in range(m):
        for j in range(n):
            linha_obj.append(-float(custos[i][j]))
    
    linha_obj += [0.0] * (m + n)
    linha_obj.append(0.0)
    tabela.append(linha_obj)
    
    return tabela

def extrair_solucao(tabela, m, n):
    """Extrai a solução da tabela simplex final"""
    num_vars = m * n
    valores = [0.0] * num_vars
    
    for j in range(num_vars):
        coluna = [tabela[i][j] for i in range(len(tabela) - 1)]
        
        if coluna.count(1) == 1 and coluna.count(0) == len(coluna) - 1:
            linha_base = coluna.index(1)
            valores[j] = max(0, tabela[linha_base][-1])
    
    custo_total = -tabela[-1][-1]
    
    return valores, custo_total

def gerar_problema_transporte(m, n, total=100000, semente=42):
    """Gera um problema de transporte balanceado"""
    random.seed(semente)
    
    # Gerar oferta
    oferta = [random.randint(1000, 4000) for _ in range(m)]
    escala = total / sum(oferta)
    oferta = [int(x * escala) for x in oferta]
    oferta[-1] += total - sum(oferta)
    
    # Gerar demanda
    demanda = [random.randint(1000, 4000) for _ in range(n)]
    escala = total / sum(demanda)
    demanda = [int(x * escala) for x in demanda]
    demanda[-1] += total - sum(demanda)
    
    # Gerar custos
    custos = [[random.randint(1, 100) for _ in range(n)] for _ in range(m)]
    
    return oferta, demanda, custos

def executar_benchmark(m, n, num_repeticoes=10):
    """Executa benchmark para um tamanho específico"""
    print(f"\n{'='*60}")
    print(f"BENCHMARK: {m}×{n} - {num_repeticoes} repetições")
    print(f"{'='*60}")
    
    resultados = {
        'tamanho': f"{m}x{n}",
        'm': m,
        'n': n,
        'num_repeticoes': num_repeticoes,
        'execucoes': []
    }
    
    for i in range(num_repeticoes):
        print(f"\nExecução {i+1}/{num_repeticoes}...", end=" ")
        
        # Forçar coleta de lixo antes de medir memória
        gc.collect()
        
        processo = psutil.Process(os.getpid())
        
        # Medir memória ANTES - usar working set (Windows) ou RSS (Linux/Mac)
        try:
            # Windows: usa working set
            mem_antes = processo.memory_info().wset / (1024 * 1024)
        except AttributeError:
            # Linux/Mac: usa RSS
            mem_antes = processo.memory_info().rss / (1024 * 1024)
        
        # Gerar problema
        tempo_inicio_total = time.time()
        oferta, demanda, custos = gerar_problema_transporte(m, n, semente=42+i)
        
        # Construir tabela
        tempo_inicio_construcao = time.time()
        tabela = construir_tabela_transporte(oferta, demanda, custos)
        tempo_construcao = time.time() - tempo_inicio_construcao
        
        # Resolver
        tempo_inicio_simplex = time.time()
        iteracoes = simplex(tabela, verbose=False)
        tempo_simplex = time.time() - tempo_inicio_simplex
        
        # Extrair solução
        valores, custo_total = extrair_solucao(tabela, m, n)
        
        tempo_total = time.time() - tempo_inicio_total
        
        # Medir memória DEPOIS
        try:
            # Windows: usa working set
            mem_depois = processo.memory_info().wset / (1024 * 1024)
        except AttributeError:
            # Linux/Mac: usa RSS
            mem_depois = processo.memory_info().rss / (1024 * 1024)
        
        # Calcular diferença - garantir que seja positiva ou zero
        memoria_usada = max(0, mem_depois - mem_antes)
        
        # Se a memória for muito pequena ou negativa, usar memória total atual
        # (acontece quando o garbage collector libera memória durante a execução)
        if memoria_usada < 0.1:
            try:
                memoria_usada = processo.memory_info().wset / (1024 * 1024)
            except AttributeError:
                memoria_usada = processo.memory_info().rss / (1024 * 1024)
        
        exec_resultado = {
            'execucao': i + 1,
            'tempo_total': tempo_total,
            'tempo_construcao': tempo_construcao,
            'tempo_simplex': tempo_simplex,
            'memoria_mb': memoria_usada,
            'memoria_antes_mb': mem_antes,
            'memoria_depois_mb': mem_depois,
            'iteracoes': iteracoes,
            'custo_total': custo_total
        }
        
        resultados['execucoes'].append(exec_resultado)
        print(f"OK - {tempo_total:.4f}s - {iteracoes} iterações - {memoria_usada:.2f} MB")
        
        # Limpar memória após cada execução
        del tabela, valores, oferta, demanda, custos
        gc.collect()
    
    # Calcular estatísticas
    tempos = [e['tempo_total'] for e in resultados['execucoes']]
    tempos_simplex = [e['tempo_simplex'] for e in resultados['execucoes']]
    memorias = [e['memoria_mb'] for e in resultados['execucoes']]
    iteracoes_list = [e['iteracoes'] for e in resultados['execucoes']]
    
    resultados['estatisticas'] = {
        'tempo_medio': statistics.mean(tempos),
        'tempo_mediano': statistics.median(tempos),
        'tempo_desvio': statistics.stdev(tempos) if len(tempos) > 1 else 0,
        'tempo_min': min(tempos),
        'tempo_max': max(tempos),
        
        'tempo_simplex_medio': statistics.mean(tempos_simplex),
        'tempo_simplex_mediano': statistics.median(tempos_simplex),
        
        'memoria_media': statistics.mean(memorias),
        'memoria_mediana': statistics.median(memorias),
        'memoria_max': max(memorias),
        'memoria_min': min(memorias),
        
        'iteracoes_media': statistics.mean(iteracoes_list),
        'iteracoes_min': min(iteracoes_list),
        'iteracoes_max': max(iteracoes_list)
    }
    
    # Mostrar resumo
    print(f"\n{'='*60}")
    print("ESTATÍSTICAS:")
    print(f"Tempo médio: {resultados['estatisticas']['tempo_medio']:.4f}s ± {resultados['estatisticas']['tempo_desvio']:.4f}s")
    print(f"Tempo Simplex médio: {resultados['estatisticas']['tempo_simplex_medio']:.4f}s")
    print(f"Memória média: {resultados['estatisticas']['memoria_media']:.2f} MB")
    print(f"Memória mediana: {resultados['estatisticas']['memoria_mediana']:.2f} MB")
    print(f"Iterações médias: {resultados['estatisticas']['iteracoes_media']:.0f}")
    print(f"{'='*60}")
    
    return resultados

# ======================
# EXECUÇÃO PRINCIPAL
# ======================

if __name__ == "__main__":
    print("="*60)
    print("BENCHMARK SIMPLEX - PYTHON (Versão Corrigida)")
    print("="*60)
    
    # Informações do sistema
    processo = psutil.Process(os.getpid())
    print(f"\nSistema Operacional: {os.name}")
    print(f"CPUs disponíveis: {psutil.cpu_count()}")
    print(f"Memória RAM total: {psutil.virtual_memory().total / (1024**3):.2f} GB")
    
    # Configurações de teste
    tamanhos = [
        (50, 50),
        (60, 60),
        (70, 70),
        (80, 80),
        (90, 90),
        (100, 100)
    ]
    
    num_repeticoes = 10
    
    todos_resultados = []
    
    for m, n in tamanhos:
        resultado = executar_benchmark(m, n, num_repeticoes)
        todos_resultados.append(resultado)
    
    # Salvar resultados em JSON
    timestamp = time.strftime("%Y%m%d_%H%M%S")
    nome_arquivo = f"benchmark_python_{timestamp}.json"
    
    with open(nome_arquivo, 'w', encoding='utf-8') as f:
        json.dump(todos_resultados, f, indent=2, ensure_ascii=False)
    
    print(f"\n{'='*60}")
    print(f"Resultados salvos em: {nome_arquivo}")
    print(f"{'='*60}")
    
    # Resumo final
    print("\n" + "="*60)
    print("RESUMO GERAL:")
    print("="*60)
    print(f"{'Tamanho':<12} {'Tempo Médio':<15} {'Iterações':<12} {'Memória (MB)'}")
    print("-"*60)
    for r in todos_resultados:
        print(f"{r['tamanho']:<12} {r['estatisticas']['tempo_medio']:<15.4f} "
              f"{r['estatisticas']['iteracoes_media']:<12.0f} "
              f"{r['estatisticas']['memoria_media']:.2f}")
    print("="*60)

File: C/benchmark_simplex.c
#include <stdio.h>
#include <stdlib.h>
#include <float.h>
#include <time.h>
#include <math.h>
#include <sys/time.h>
#include <sys/resource.h>
#include <unistd.h>
#include <string.h>
#include <limits.h>

typedef struct {
    int m;
    int n;
    int execucao;
    double tempo_total;
    double tempo_construcao;
    double tempo_simplex;
    double memoria_mb;
    int iteracoes;
    double custo_total;
} ResultadoExecucao;

typedef struct {
    char tamanho[20];
    int m;
    int n;
    int num_repeticoes;
    ResultadoExecucao *execucoes;
    double tempo_medio;
    double tempo_mediano;
    double tempo_desvio;
    double tempo_min;
    double tempo_max;
    double tempo_simplex_medio;
    double memoria_media;
    double memoria_max;
    double iteracoes_media;
    int iteracoes_min;
    int iteracoes_max;
} ResultadoBenchmark;

// Função para obter tempo em segundos com alta precisão
double obter_tempo() {
    struct timeval tv;
    gettimeofday(&tv, NULL);
    return tv.tv_sec + tv.tv_usec / 1000000.0;
}

// Função para obter memória usada em MB
double memoria_usada_MB() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    // No Linux, ru_maxrss está em kilobytes
    return usage.ru_maxrss / 1024.0;
}

// Alocar matriz 2D
double** alocar_matriz(int linhas, int colunas) {
    double **matriz = (double**)malloc(linhas * sizeof(double*));
    for (int i = 0; i < linhas; i++) {
        matriz[i] = (double*)calloc(colunas, sizeof(double));
    }
    return matriz;
}

// Liberar matriz 2D
void liberar_matriz(double **matriz, int linhas) {
    for (int i = 0; i < linhas; i++) {
        free(matriz[i]);
    }
    free(matriz);
}

// Encontrar coluna pivô
int encontrar_coluna_pivo(double **tabela, int linhas, int colunas) {
    double valor_min = 0;
    int coluna_pivo = -1;
    
    for (int j = 0; j < colunas - 1; j++) {
        if (tabela[linhas - 1][j] < valor_min) {
            valor_min = tabela[linhas - 1][j];
            coluna_pivo = j;
        }
    }
    return coluna_pivo;
}

// Encontrar linha pivô
int encontrar_linha_pivo(double **tabela, int linhas, int colunas, int coluna_pivo) {
    double menor = DBL_MAX;
    int linha_pivo = -1;
    
    for (int i = 0; i < linhas - 1; i++) {
        double elemento = tabela[i][coluna_pivo];
        if (elemento > 0) {
            double razao = tabela[i][colunas - 1] / elemento;
            if (razao < menor) {
                menor = razao;
                linha_pivo = i;
            }
        }
    }
    return linha_pivo;
}

// Pivotear
void pivotear(double **tabela, int linhas, int colunas, int linha_pivo, int coluna_pivo) {
    double pivo = tabela[linha_pivo][coluna_pivo];
    
    for (int j = 0; j < colunas; j++) {
        tabela[linha_pivo][j] /= pivo;
    }
    
    for (int i = 0; i < linhas; i++) {
        if (i != linha_pivo) {
            double multiplicador = tabela[i][coluna_pivo];
            for (int j = 0; j < colunas; j++) {
                tabela[i][j] -= multiplicador * tabela[linha_pivo][j];
            }
        }
    }
}

// Algoritmo Simplex - retorna número de iterações
int simplex(double **tabela, int linhas, int colunas, int max_iter) {
    int iteracao = 0;
    
    while (iteracao < max_iter) {
        iteracao++;
        
        int coluna_pivo = encontrar_coluna_pivo(tabela, linhas, colunas);
        if (coluna_pivo == -1) {
            return iteracao;
        }
        
        int linha_pivo = encontrar_linha_pivo(tabela, linhas, colunas, coluna_pivo);
        if (linha_pivo == -1) {
            printf("Problema ilimitado.\n");
            return -1;
        }
        
        pivotear(tabela, linhas, colunas, linha_pivo, coluna_pivo);
    }
    
    printf("ATENÇÃO: Limite de %d iterações atingido!\n", max_iter);
    return iteracao;
}

// Construir tabela simplex
double** construir_tabela_transporte(int *oferta, int *demanda, int **custos, 
                                     int m, int n, int *linhas_ret, int *colunas_ret) {
    int num_vars = m * n;
    int total_linhas = m + n + 1;
    int total_colunas = num_vars + m + n + 1;
    
    *linhas_ret = total_linhas;
    *colunas_ret = total_colunas;
    
    double **tabela = alocar_matriz(total_linhas, total_colunas);
    
    // Restrições de oferta
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            tabela[i][i * n + j] = 1;
        }
        tabela[i][num_vars + i] = 1;
        tabela[i][total_colunas - 1] = oferta[i];
    }
    
    // Restrições de demanda
    for (int j = 0; j < n; j++) {
        for (int i = 0; i < m; i++) {
            tabela[m + j][i * n + j] = 1;
        }
        tabela[m + j][num_vars + m + j] = 1;
        tabela[m + j][total_colunas - 1] = demanda[j];
    }
    
    // Função objetivo
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            tabela[total_linhas - 1][i * n + j] = -custos[i][j];
        }
    }
    
    return tabela;
}

// Extrair solução
double extrair_solucao(double **tabela, int linhas, int colunas, int m, int n) {
    return -tabela[linhas - 1][colunas - 1];
}

// Gerar problema de transporte
void gerar_problema_transporte(int m, int n, int **oferta_ret, int **demanda_ret, 
                               int ***custos_ret, int semente) {
    int total = 100000;
    
    int *oferta = (int*)malloc(m * sizeof(int));
    int *demanda = (int*)malloc(n * sizeof(int));
    int **custos = (int**)malloc(m * sizeof(int*));
    
    srand(semente);
    
    // Gerar oferta
    int soma_oferta = 0;
    for (int i = 0; i < m; i++) {
        oferta[i] = rand() % 3000 + 1000;
        soma_oferta += oferta[i];
    }
    
    double escala = (double)total / soma_oferta;
    soma_oferta = 0;
    for (int i = 0; i < m; i++) {
        oferta[i] = (int)(oferta[i] * escala);
        soma_oferta += oferta[i];
    }
    oferta[m - 1] += total - soma_oferta;
    
    // Gerar demanda
    int soma_demanda = 0;
    for (int j = 0; j < n; j++) {
        demanda[j] = rand() % 3000 + 1000;
        soma_demanda += demanda[j];
    }
    
    escala = (double)total / soma_demanda;
    soma_demanda = 0;
    for (int j = 0; j < n; j++) {
        demanda[j] = (int)(demanda[j] * escala);
        soma_demanda += demanda[j];
    }
    demanda[n - 1] += total - soma_demanda;
    
    // Gerar custos
    for (int i = 0; i < m; i++) {
        custos[i] = (int*)malloc(n * sizeof(int));
        for (int j = 0; j < n; j++) {
            custos[i][j] = rand() % 100 + 1;
        }
    }
    
    *oferta_ret = oferta;
    *demanda_ret = demanda;
    *custos_ret = custos;
}

// Função de comparação para qsort
int comparar_double(const void *a, const void *b) {
    double diff = *(double*)a - *(double*)b;
    return (diff > 0) - (diff < 0);
}

// Calcular estatísticas
void calcular_estatisticas(ResultadoBenchmark *resultado) {
    int n = resultado->num_repeticoes;
    
    double soma_tempo = 0, soma_tempo_simplex = 0, soma_memoria = 0;
    double soma_iteracoes = 0;
    
    double *tempos = (double*)malloc(n * sizeof(double));
    
    resultado->tempo_min = DBL_MAX;
    resultado->tempo_max = 0;
    resultado->iteracoes_min = INT_MAX;
    resultado->iteracoes_max = 0;
    resultado->memoria_max = 0;
    
    for (int i = 0; i < n; i++) {
        ResultadoExecucao *exec = &resultado->execucoes[i];
        
        tempos[i] = exec->tempo_total;
        soma_tempo += exec->tempo_total;
        soma_tempo_simplex += exec->tempo_simplex;
        soma_memoria += exec->memoria_mb;
        soma_iteracoes += exec->iteracoes;
        
        if (exec->tempo_total < resultado->tempo_min) resultado->tempo_min = exec->tempo_total;
        if (exec->tempo_total > resultado->tempo_max) resultado->tempo_max = exec->tempo_total;
        if (exec->iteracoes < resultado->iteracoes_min) resultado->iteracoes_min = exec->iteracoes;
        if (exec->iteracoes > resultado->iteracoes_max) resultado->iteracoes_max = exec->iteracoes;
        if (exec->memoria_mb > resultado->memoria_max) resultado->memoria_max = exec->memoria_mb;
    }
    
    resultado->tempo_medio = soma_tempo / n;
    resultado->tempo_simplex_medio = soma_tempo_simplex / n;
    resultado->memoria_media = soma_memoria / n;
    resultado->iteracoes_media = soma_iteracoes / n;
    
    // Calcular mediana
    qsort(tempos, n, sizeof(double), comparar_double);
    if (n % 2 == 0) {
        resultado->tempo_mediano = (tempos[n/2-1] + tempos[n/2]) / 2.0;
    } else {
        resultado->tempo_mediano = tempos[n/2];
    }
    
    // Calcular desvio padrão
    double soma_quad = 0;
    for (int i = 0; i < n; i++) {
        double diff = resultado->execucoes[i].tempo_total - resultado->tempo_medio;
        soma_quad += diff * diff;
    }
    resultado->tempo_desvio = sqrt(soma_quad / n);
    
    free(tempos);
}

// Executar benchmark
ResultadoBenchmark executar_benchmark(int m, int n, int num_repeticoes) {
    printf("\n============================================================\n");
    printf("BENCHMARK: %dx%d - %d repeticoes\n", m, n, num_repeticoes);
    printf("============================================================\n");
    
    ResultadoBenchmark resultado;
    sprintf(resultado.tamanho, "%dx%d", m, n);
    resultado.m = m;
    resultado.n = n;
    resultado.num_repeticoes = num_repeticoes;
    resultado.execucoes = (ResultadoExecucao*)malloc(num_repeticoes * sizeof(ResultadoExecucao));
    
    for (int i = 0; i < num_repeticoes; i++) {
        printf("\nExecucao %d/%d... ", i+1, num_repeticoes);
        fflush(stdout);
        
        double mem_antes = memoria_usada_MB();
        double inicio_total = obter_tempo();
        
        // Gerar problema
        int *oferta, *demanda, **custos;
        gerar_problema_transporte(m, n, &oferta, &demanda, &custos, 42 + i);
        
        // Construir tabela
        double inicio_construcao = obter_tempo();
        int linhas, colunas;
        double **tabela = construir_tabela_transporte(oferta, demanda, custos, m, n, &linhas, &colunas);
        double fim_construcao = obter_tempo();
        
        // Resolver
        double inicio_simplex = obter_tempo();
        int iteracoes = simplex(tabela, linhas, colunas, 1000000);
        double fim_simplex = obter_tempo();
        
        // Extrair solução
        double custo_total = extrair_solucao(tabela, linhas, colunas, m, n);
        
        double fim_total = obter_tempo();
        double mem_depois = memoria_usada_MB();
        
        // Salvar resultados
        ResultadoExecucao *exec = &resultado.execucoes[i];
        exec->m = m;
        exec->n = n;
        exec->execucao = i + 1;
        exec->tempo_total = fim_total - inicio_total;
        exec->tempo_construcao = fim_construcao - inicio_construcao;
        exec->tempo_simplex = fim_simplex - inicio_simplex;
        exec->memoria_mb = mem_depois - mem_antes;
        exec->iteracoes = iteracoes;
        exec->custo_total = custo_total;
        
        printf("OK - %.4fs - %d iteracoes", exec->tempo_total, iteracoes);
        
        // Liberar memória
        liberar_matriz(tabela, linhas);
        free(oferta);
        free(demanda);
        for (int j = 0; j < m; j++) {
            free(custos[j]);
        }
        free(custos);
    }
    
    // Calcular estatísticas
    calcular_estatisticas(&resultado);
    
    // Mostrar resumo
    printf("\n============================================================\n");
    printf("ESTATISTICAS:\n");
    printf("Tempo medio: %.4fs +/- %.4fs\n", resultado.tempo_medio, resultado.tempo_desvio);
    printf("Tempo Simplex medio: %.4fs\n", resultado.tempo_simplex_medio);
    printf("Memoria media: %.2f MB\n", resultado.memoria_media);
    printf("Iteracoes medias: %.0f\n", resultado.iteracoes_media);
    printf("============================================================\n");
    
    return resultado;
}

// Salvar resultados em arquivo JSON
void salvar_resultados_json(ResultadoBenchmark *resultados, int num_testes, const char *nome_arquivo) {
    FILE *fp = fopen(nome_arquivo, "w");
    if (fp == NULL) {
        printf("Erro ao criar arquivo JSON!\n");
        return;
    }
    
    fprintf(fp, "[\n");
    
    for (int t = 0; t < num_testes; t++) {
        ResultadoBenchmark *r = &resultados[t];
        
        fprintf(fp, "  {\n");
        fprintf(fp, "    \"tamanho\": \"%s\",\n", r->tamanho);
        fprintf(fp, "    \"m\": %d,\n", r->m);
        fprintf(fp, "    \"n\": %d,\n", r->n);
        fprintf(fp, "    \"num_repeticoes\": %d,\n", r->num_repeticoes);
        fprintf(fp, "    \"execucoes\": [\n");
        
        for (int i = 0; i < r->num_repeticoes; i++) {
            ResultadoExecucao *e = &r->execucoes[i];
            fprintf(fp, "      {\n");
            fprintf(fp, "        \"execucao\": %d,\n", e->execucao);
            fprintf(fp, "        \"tempo_total\": %.6f,\n", e->tempo_total);
            fprintf(fp, "        \"tempo_construcao\": %.6f,\n", e->tempo_construcao);
            fprintf(fp, "        \"tempo_simplex\": %.6f,\n", e->tempo_simplex);
            fprintf(fp, "        \"memoria_mb\": %.2f,\n", e->memoria_mb);
            fprintf(fp, "        \"iteracoes\": %d,\n", e->iteracoes);
            fprintf(fp, "        \"custo_total\": %.2f\n", e->custo_total);
            fprintf(fp, "      }%s\n", (i < r->num_repeticoes - 1) ? "," : "");
        }
        
        fprintf(fp, "    ],\n");
        fprintf(fp, "    \"estatisticas\": {\n");
        fprintf(fp, "      \"tempo_medio\": %.6f,\n", r->tempo_medio);
        fprintf(fp, "      \"tempo_mediano\": %.6f,\n", r->tempo_mediano);
        fprintf(fp, "      \"tempo_desvio\": %.6f,\n", r->tempo_desvio);
        fprintf(fp, "      \"tempo_min\": %.6f,\n", r->tempo_min);
        fprintf(fp, "      \"tempo_max\": %.6f,\n", r->tempo_max);
        fprintf(fp, "      \"tempo_simplex_medio\": %.6f,\n", r->tempo_simplex_medio);
        fprintf(fp, "      \"memoria_media\": %.2f,\n", r->memoria_media);
        fprintf(fp, "      \"memoria_max\": %.2f,\n", r->memoria_max);
        fprintf(fp, "      \"iteracoes_media\": %.0f,\n", r->iteracoes_media);
        fprintf(fp, "      \"iteracoes_min\": %d,\n", r->iteracoes_min);
        fprintf(fp, "      \"iteracoes_max\": %d\n", r->iteracoes_max);
        fprintf(fp, "    }\n");
        fprintf(fp, "  }%s\n", (t < num_testes - 1) ? "," : "");
    }
    
    fprintf(fp, "]\n");
    fclose(fp);
    
    printf("\nResultados salvos em: %s\n", nome_arquivo);
}

// Salvar resultados em CSV
void salvar_resultados_csv(ResultadoBenchmark *resultados, int num_testes, const char *nome_arquivo) {
    FILE *fp = fopen(nome_arquivo, "w");
    if (fp == NULL) {
        printf("Erro ao criar arquivo CSV!\n");
        return;
    }
    
    // Cabeçalho
    fprintf(fp, "Tamanho,M,N,Execucao,Tempo_Total,Tempo_Construcao,Tempo_Simplex,Memoria_MB,Iteracoes,Custo_Total\n");
    
    // Dados
    for (int t = 0; t < num_testes; t++) {
        ResultadoBenchmark *r = &resultados[t];
        for (int i = 0; i < r->num_repeticoes; i++) {
            ResultadoExecucao *e = &r->execucoes[i];
            fprintf(fp, "%s,%d,%d,%d,%.6f,%.6f,%.6f,%.2f,%d,%.2f\n",
                    r->tamanho, r->m, r->n, e->execucao,
                    e->tempo_total, e->tempo_construcao, e->tempo_simplex,
                    e->memoria_mb, e->iteracoes, e->custo_total);
        }
    }
    
    fclose(fp);
    printf("Resultados salvos em: %s\n", nome_arquivo);
}

// ======================
// FUNÇÃO PRINCIPAL
// ======================
int main() {
    printf("============================================================\n");
    printf("BENCHMARK SIMPLEX - LINUX\n");
    printf("============================================================\n");
    
    // Configurações de teste
    int tamanhos[][2] = {
        {5, 5},
        {10, 10},
        {15, 15},
        {20, 20},
        {25, 25},
        {30, 30}
    };
    
    int num_testes = sizeof(tamanhos) / sizeof(tamanhos[0]);
    int num_repeticoes = 10;
    
    ResultadoBenchmark *todos_resultados = (ResultadoBenchmark*)malloc(num_testes * sizeof(ResultadoBenchmark));
    
    // Executar todos os benchmarks
    for (int i = 0; i < num_testes; i++) {
        todos_resultados[i] = executar_benchmark(tamanhos[i][0], tamanhos[i][1], num_repeticoes);
    }
    
    // Gerar nome de arquivo com timestamp
    time_t t = time(NULL);
    struct tm *tm_info = localtime(&t);
    char timestamp[20];
    strftime(timestamp, sizeof(timestamp), "%Y%m%d_%H%M%S", tm_info);
    
    char nome_json[100], nome_csv[100];
    sprintf(nome_json, "benchmark_linux_%s.json", timestamp);
    sprintf(nome_csv, "benchmark_linux_%s.csv", timestamp);
    
    // Salvar resultados
    salvar_resultados_json(todos_resultados, num_testes, nome_json);
    salvar_resultados_csv(todos_resultados, num_testes, nome_csv);
    
    // Resumo final
    printf("\n============================================================\n");
    printf("RESUMO GERAL:\n");
    printf("============================================================\n");
    printf("%-12s %-15s %-12s %-15s\n", "Tamanho", "Tempo Medio", "Iteracoes", "Memoria (MB)");
    printf("------------------------------------------------------------\n");
    
    for (int i = 0; i < num_testes; i++) {
        ResultadoBenchmark *r = &todos_resultados[i];
        printf("%-12s %-15.4f %-12.0f %-15.2f\n",
               r->tamanho, r->tempo_medio, r->iteracoes_media, r->memoria_media);
    }
    
    printf("============================================================\n");
    
    // Liberar memória
    for (int i = 0; i < num_testes; i++) {
        free(todos_resultados[i].execucoes);
    }
    free(todos_resultados);
    
    return 0;
}